Refactor tests, fix errors shown [changelog skip]

MSP-Greg opened the pull request:
1. Clean up leaked fd's both in test & lib code

2. Delete control io's and unix paths

3. Fix code differences between signal and control_cli behavior

4. control_cli.rb - refactor, add thread-status using sock control

5. launcher.rb - add thread-status via sock control

6. detect.rb - Add Puma::HAS_FORK and Puma::HAS_UNIX, adjust tests to match

7. Fixed connection handling during stop (both Single & Clustered)

8. split some test files into two classes, one running parallel, one 'serial'

MSP-Greg commented:
This is a lot of code.  It started when I was working on tests, noticed odd things happening, and decided to write some code to check leaked fd's.  There were quite a few.  Some were from the tests, some were from the code.  The test code will dump fd's on test failures and at the end of testing if there are more than three open fd's (STDOUT, STDERR, & STDIN).

Also, tried to minimize the 'test timed out' issues by cleaning up some of the code that was testing io read values, etc.

While working on test coverage, I was too lazy to shut off Actions, as the coverage was running in Travis.  This has passed a gazillion builds without failure, and a minimum number re-tried tests.

Some of the code updates the 'integration' tests, which are typically 'front to back' tests, but many utilize signals, which means the Puma instance must be in a separate process.  Another reason is 'restart', which has `Kernel.exec` calls, which don't work to well in threads...

Several of the 'helper methods' help with things like closing sockets, etc, which are easy to overlook when writing tests.  I started documenting the 'helper' methods and classes, but haven't finished.  These could be included in normal Puma docs, kind of like RubyGems TestCase.

One thing that should be added is binding to an ssl socket, but I thought this was far enough along for a draft PR.

Similar code could/should be added to tests that are running a Puma instance in a thread, as several tests do that.  That can wait until another PR.

Lastly, with large problems, I often do not work in a linear fashion, but I seem to converge on a solution.  Happy to divide this up, but need suggestions as to how...

nateberkopec commented:
Will look more later but I don’t like running tests on all branches for pushes. You end up with two builds when making PRs from this remote.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

MSP-Greg commented:
I removed the change re branches and the additional rake file.

nateberkopec commented on lib/puma/app/status.rb:
```diff
@@ -16,47 +16,51 @@ def initialize(cli, token = nil)
+            rack_response 404, '{ "error": "reload_worker_directory not available" }'
           else
-            rack_response(200, OK_STATUS)
+            rack_response 200, OK_STATUS
```
If we're gonna do this (also changing to single-quotes), we should also be added Rubocop rules.

MSP-Greg replied:
BTW, I didn't change all of them (currently in master), note both single & double quotes.  So I just made them consistent...

https://github.com/puma/puma/blob/1d97e42dde2bcc737f8d01f6c2292f6845c9c504/lib/puma/app/status.rb#L59

nateberkopec commented on lib/puma/control_cli.rb:
```diff
@@ -191,42 +204,22 @@ def send_signal
-
-          return
-
+        if sgnl = CMDS_SGNLS[@command]
```
Mistake? `==`.

nateberkopec commented:
I'm conceptually OK with all of this, but I think we should be adding Rubocop rules any time we're making style changes.

MSP-Greg commented:
I'm sorry for the style changes.  I can remove them tomorrow/today.  Been looking at monitors since before Apple II days, and I'm 60+.  After a long day, this:
```ruby
rack_response 200, OK_STATUS
```
is easier to read than:
```ruby
rack_response(200, OK_STATUS)
```
Re RuboCop, I've seen some of the discussions in ruby/ruby about it, and it's a pretty solid NO there...

JFYI, this makes JRuby more stable, but stable in the sense that it consistently fails, but seems to more often get to the test summary.  I tried, I give up.  Someone familiar with JRuby will have to look at it.  It's rock solid on Ubuntu & macOS.

Also, I've got inherited tests (tcp, unix & ssl) working in TestBinder.  Much better coverage.  Setup to go on top of this.

MSP-Greg commented:
Just rebased on master, should be passing on Windows...

nateberkopec commented on test/test_pumactl.rb:
```diff
@@ -106,50 +97,49 @@ def test_default_config_file_exist
+    assert_io 'Goodbye', io: @wait
+
+    #Puma::ControlCLI.new(opts + ['halt'], @ready, @ready).run
+    #assert_io 'Stopping immediately!', io: @wait, debug: true
```
Should this still be commented?

MSP-Greg replied:
IDK.  You may have seen errors where Events is trying to write but the pipe/IO is closed?  Those were part of what I was trying to clean up, so I may have switched to stop early in the work on it.

Also, IDK if we've got halt covered in all of the four classes that include it.  I'll look at tomorrow.

This PR was meant as a start to improving the test framework, both in terms of stability and ease of use.  I wanted it to be trivial to do tests for the same functionality, with one using a control socket and one using a signal.  I didn't want to mess with UNIX paths, remembering to close all the IO's, have leaky fd's, etc.  Also, no more naked read statements leading to lengthy 'test took too long' timeouts when a good method should limit them to 5 or 10 seconds, etc, etc..

I did have to use it in enough test files to see if it was functional, and it changed/expanded as I was developing it.  So, I'm sure this will change in the future, but I think it's a good first step.  

And, given how far I got into things with the Binder work and the systemd stuff, I feel pretty confident about the changes in the lib files.

Sorry for the extended post, I've wrestled with some messy things today.  There are a few small changes I'll push over the weekend.

MSP-Greg replied:
@nateberkopec

Hate to ask, can you try the above commented 'halt' lines with the two 'stop' lines commented?  Locally it freezes, probably why I commented it (a reminder for later).  I can push on my fork to Actions also...

MSP-Greg replied:
I rebased and deleted the comments.  We should see halt no being called in test coverage.

MSP-Greg commented:
One thing not explicitly mentioned in the above comments is that current behavior shows reset connections during `term_closes_listeners` tcp tests.  The tests check connection handling after a stop command is issued.  I think it was discussed when the debug output was added to test logging.  This PR adds more tests for this behavior.

This PR also fixes the 'reset' issue.  Connections are either 'successes' or 'refused', we no longer have 'resets'...

MSP-Greg commented:
Due to intermittent (but okay) results in the `stop_closes_listeners` tests, one more change.

The tests throw 40 connections at a server being stopped, 10 before, 30 after.  On macOS, occasionally 11 connections are correctly responded to.  Other platforms it seems to always be 10.  

Adjusted asserts so the test is not run again.

see https://github.com/puma/puma/runs/268563842#step:9:288

Besides, showing that the PR is certainly stable.  Sorry, the one time I didn't run rubocop first...

MSP-Greg commented:
Thanks for reminding me about that comment.  Fixed `halt`, added a test for anytime the integration tests use a UNIX socket for control that Puma deletes the file, and added several `halt` tests.

With the changes, this is done from my perspective.

nateberkopec commented on lib/puma/launcher.rb:
```diff
@@ -184,7 +185,9 @@ def run
       end
-      @binder.close_unix_paths
+        @binder.close
+        @binder.close_unix_paths
```
I just realized - do we really want to close all sockets when restarting? Isn't that exactly the opposite of what we're supposed to do?

MSP-Greg replied:
I don't think the code ever reaches here in restarts.  It hits the Kernel.exec statements.

nateberkopec commented on lib/puma/launcher.rb:
```diff
@@ -195,7 +198,7 @@ def connected_port
       cmd = @options[:restart_cmd]
       if cmd
-        cmd.split(' ') + @original_argv
+        cmd.split + @original_argv
```
I don't think we should remove this, as it depends on `$;` being nil.

MSP-Greg replied:
I've seen that a lot of places, including Ruby core, but I'll change it.

nateberkopec commented on lib/puma/reactor.rb:
```diff
@@ -244,7 +244,8 @@ def run_internal
 
-                c.close
+                # may generate another SSL error, continue to remove the monitor
+                c.close rescue nil
```
Was this in response to a specific issue?

MSP-Greg replied:
Yes.  Where, don't recall.  I can comment it out and run the tests.

nateberkopec commented on lib/puma/reactor.rb:
```diff
@@ -302,6 +303,7 @@ def clear_monitor(mon)
     def run
       run_internal
     ensure
+      @selector.close
```
Same here, was this in response to a specific issue?

MSP-Greg replied:
nio4r doc for the method: 'Close the selector and free system resources'

I don't recall whether it affected leaky fd's or not.  It should be done somewhere...

nateberkopec commented on lib/puma/server.rb:
```diff
@@ -258,6 +258,7 @@ def handle_servers_lopez_mode
         STDERR.puts e.backtrace
       ensure
         begin
+          @notify.close rescue nil
```
Makes sense, seems like an oversight. Is this tested?

MSP-Greg replied:
The test suite did not previously test for leaky fd's.  Right or wrong, I decided that doing a `GC.start` & `ObjectSpace.each_object(IO)` for most tests might set the CI time out too far.  So I just did it when the tests finished, and made it easy to check individual tests if needed.

When I first added it, the fd's were well over 100.

Summing up, there are no asserts for leaky fd's at present, just the check at the end.

nateberkopec commented on lib/puma/server.rb:
```diff
@@ -428,8 +429,8 @@ def handle_servers
       ensure
-        @check.close
-        @notify.close
+        @check.close  rescue nil
```
What was the motivation for adding rescue nil?

MSP-Greg replied:
Exactly, I don't recall.  We're in an ensure clause at shutdown ?

nateberkopec commented on lib/puma/single.rb:
```diff
@@ -30,13 +30,14 @@ def stop
 
     def halt
-      @server.halt
+      @server.halt if @server
```
How does `@server` become nil?

MSP-Greg replied:
Good question.  I saw it used elsewhere, and again, this is 'exit' code.

nateberkopec commented on lib/puma/runner.rb:
```diff
@@ -88,6 +88,20 @@ def start_control
+      end
+    end
+
+    def close_control_io
```
The control server itself should just handle this. I can't think of a circumstance where we want to stop the control server but want to keep its IOs open, right?

nateberkopec replied:
i.e. no one should call close_control_io, everywhere should call end_control(type).

MSP-Greg replied:
> The control server itself should just handle this

The problem is that there isn't a control server class, and servers can also inherit binders.  So closing anything in the `Server` class is dicey.  Won't make that mistake again...

Re `close_control_io`, I think I can get rid of it, I'll need to add a `sync` arg to the `end_control` method.

nateberkopec commented on test/helper.rb:
```diff
@@ -45,6 +49,35 @@ def hit(uris)
+
+    ttf = Time.now.to_f + 10
+
+    while IO.select([io], nil, nil, timeout) do
```
I feel like we could use more descriptive variable names in this method, I think you slipped into writing C here 😆

MSP-Greg replied:
Yeah, I started writing that while frustrated.

Question: I've got STDERR going to the IO.pipe output for integration tests, and many of the other tests map Events IO to one IO.

Should the method dump all the read data to the console if the assert fails?

nateberkopec commented on test/helper.rb:
```diff
@@ -79,12 +112,11 @@ module TestSkips
-  # socket is required by puma
-  # usage: skip UNIX_SKT_MSG unless UNIX_SKT_EXIST
-  UNIX_SKT_EXIST = Object.const_defined? :UNIXSocket
+  # usage: skip UNIX_SKT_MSG unless HAS_UNIXHAS_UNIX
```
typo at EOL

MSP-Greg replied:
I knew that...  Thanks,

nateberkopec commented on test/helper.rb:
```diff
@@ -125,6 +157,31 @@ def skip_unless(eng, bt: caller)
     skip skip_msg, bt if skip_msg
   end
+
+  def out_io
```
This is gonna need a comment. It looks like you're trying to make sure all ios are closed after the suite ends, right?

In that case, I think this needs assertions (not just debug output) and instead of returning against length, it would be much more descriptive if `ary` contained STDOUT/STDERR/whatever the third thing is.

MSP-Greg replied:
First: Correct.

The third thing (STDIN)?  Sorry.  I've never checked into whether one can fail in `Minitest.after_run`....

nateberkopec commented on test/helper.rb:
```diff
@@ -135,18 +192,35 @@ def self.run(reporter, options = {}) # :nodoc:
 
+  def after_teardown
+    return if passed? || skipped?
+    out_io unless Puma.jruby? or RUBY_VERSION < '2.3'
```
What's different about jruby and Ruby 2.2?

MSP-Greg replied:
I'll double check.  Something must have happened, don't recall what.

nateberkopec commented on test/helper.rb:
```diff
@@ -135,18 +192,35 @@ def self.run(reporter, options = {}) # :nodoc:
+
+    out << files
+
+    unless out.strip.empty?
```
I think this just be an assertion, not debug output.

MSP-Greg replied:
Again, `Minitest.after_run` isn't in a test.  I'll check

nateberkopec commented on test/helper.rb:
```diff
@@ -135,18 +192,35 @@ def self.run(reporter, options = {}) # :nodoc:
   unless $debugging_hold
-    out = $debugging_info.strip
-    unless out.empty?
+    out = "#{$debugging_info.rstrip}\n".dup
```
Should be extracted into its own method, like `assert_no_open_tmp_files`

MSP-Greg replied:
See above

nateberkopec commented on test/test_cli.rb:
```diff
@@ -9,34 +9,31 @@ class TestCLI < Minitest::Test
 
     @events = Puma::Events.strings
-    @events.on_booted { @ready << "!" }
+    @events.on_booted { @ready.syswrite '!' }
```
What was wrong with the shovel?

MSP-Greg replied:
Sometime ago I came across info implying that mixing 'sys' & 'normal' on the same IO may cause odd issues.  Ruby version or platform specific, I don't recall, nor how long ago...

nateberkopec commented on test/test_cli.rb:
```diff
@@ -9,34 +9,31 @@ class TestCLI < Minitest::Test
 
-    @wait.close
-    @ready.close
+    @ready.close rescue nil
```
I think you can probably understand why I don't think just adding `rescue nil` everywhere is fixing anything, right? 😆

MSP-Greg replied:
It's in a test.  If something else goes wrong before it...

nateberkopec commented on test/test_integration_cluster.rb:
```diff
@@ -2,102 +2,122 @@
-    skip NO_FORK_MSG unless HAS_FORK
-    super
-  end
+  parallelize_me! unless Puma.jruby?
```
This line isn't ever run on JRuby anyway, is it?

MSP-Greg replied:
Ok, so maybe I forgot that JRuby doesn't support fork.  I did that before I decided I was tired of seeing all the skip messages and added the class end statement conditional.  I'll remove.

nateberkopec commented on test/test_integration_cluster.rb:
```diff
@@ -2,102 +2,122 @@
   def test_pre_existing_unix
-    skip UNIX_SKT_MSG unless UNIX_SKT_EXIST
+    skip UNIX_SKT_MSG unless HAS_UNIX
+    setup_puma bind: :unix, ctrl: :unix
```
This reads like you're setting up a control server and then not using it. Is that correct?

MSP-Greg replied:
stop_server_wait calls `run_pumactl 'stop'`

MSP-Greg commented on lib/puma/cluster.rb:
```diff
@@ -28,10 +28,12 @@ def initialize(cli, events)
 
     def stop_workers
       log "- Gracefully shutting down workers..."
+      @launcher.close_binder_listeners unless @restart
```
Some of these changes were related to leaked fd's, which is why I added the log output for it.  This may have been part of the fix of the macOS issue with ':reset' connections, which are now zero.  Previously they hovered around +/- 7.

MSP-Greg replied:
> I'm wondering what we were doing before that was wrong.

Remember your question about how some of the connections were being 'dumped' during a stop?  They were already accepted but then they were reset?  This fixes that, but we can't close them if a restart is being done, otherwise Binder#redirects_for_restart won't work because the sockets need to stay open.

nateberkopec commented:
I'm going to close this spike because it's too big to merge as-is. Please break this up into ~100-line chunks. Suggestions for chunks:

* Style changes
* 1 PR for each file changed in lib/, with corresponding tests
* 1 PR that adds the new helpers in tests
* 1 PR for each changed test file

MSP-Greg commented:
@nateberkopec

Most of the changes requested in your review can be found here:

https://github.com/MSP-Greg/puma/commit/a352600c0f253b5629153e4ea3aae5f6bc36baaa

Items that were left as-is I can explain if you'd like.
